elastic/kibana
# [Uptime] allow an administrator to enable and disable monitor management

## Summary

Fixes https://github.com/elastic/kibana/issues/127668

Creates a [`/enablement` ](https://github.com/elastic/kibana/pull/128223/files#diff-57fef2d8abf756a7485c4d7e00153ce4d94872c082382b6efb33471e0a57e5c3R11) route to support enabling Monitor Management.

Allows enablement for users with [specific privileges](https://github.com/elastic/kibana/pull/128223/files#diff-f9b84e2c63ef773caeed1d2dfcdf068978f6fa7c3f44c0bcf5b1d807332068cfR137-R151), ensuring that this user will have enough privileges to create an API key capable of writing back their elasticsearch cluster for the `synthetics-*` data streams appropriately. 

Adds [enable](https://github.com/elastic/kibana/pull/128223/files#diff-da3b402f3bcab8e45168d49a3d8b3000f31c29faa9fa203c7bcda93b0f14e3e9R96) toggle to Monitor Management.

Admin When Monitor Management is Disabled (Empty State)
<img width="1435" alt="Screen Shot 2022-03-25 at 1 53 49 PM" src="https://user-images.githubusercontent.com/11356435/160313861-b0651a85-ca3c-4cd7-baca-c642c2f8a961.png">

Admin When Monitor Management is Disabled (Existing monitors)
<img width="2152" alt="Screen Shot 2022-03-28 at 9 33 57 AM" src="https://user-images.githubusercontent.com/11356435/160409483-dbe97033-889e-45c9-ae9b-d59f10fe2cf0.png">

Admin When Monitor Management is Enabled
<img width="1418" alt="Screen Shot 2022-03-23 at 12 28 36 PM" src="https://user-images.githubusercontent.com/11356435/159748853-d6dfb035-6ae8-44ef-bf7c-18977c6725cf.png">

Uptime User When Monitor Management is Disabled (Empty State)
<img width="1439" alt="Screen Shot 2022-03-25 at 1 54 35 PM" src="https://user-images.githubusercontent.com/11356435/160313848-65e7af8d-913f-48ff-b4a4-02d27d508a04.png">

Uptime User When Monitor Management is Disabled (Existing Monitors)
<img width="2180" alt="Screen Shot 2022-03-28 at 9 34 41 AM" src="https://user-images.githubusercontent.com/11356435/160409440-2f5573e2-3206-4b9c-a147-e95843c1cae4.png">

Uptime User When Monitor Management is Enabled 
<img width="1404" alt="Screen Shot 2022-03-23 at 11 14 43 AM" src="https://user-images.githubusercontent.com/11356435/159732628-bd53c979-7e6a-4cbc-a19d-06b6992993b2.png">

### Testing basic user experience
1. Start Kibana with monitor management enabled, ideally with a fresh instance of es that does not yet have a synthetics API key defined. 
2. Define a role that has limited privileges to Uptime and `synthetics` indices by running the below command (or creating the role in Stack Management)
```
curl --insecure -u elastic:changeme -X PUT "https://localhost:5601/api/security/role/uptime_role" -H 'kbn-xsrf: true' -H 'Content-Type: application/json' -d'
{
  "metadata" : {
    "version" : 1
  },
  "elasticsearch": {
    "indices" : [
      {
        "names": [ "synthetics*" ],
        "privileges": [ "all" ]
      }
    ]
  },
  "kibana": [
    {
      "base": [],
      "feature": {
        "uptime": [
          "all"
        ]
      },
      "spaces": [
        "*"
      ]
    }
  ]
}
```
3. Define a user (our custom-created) uptime_role
```
curl -X POST -u elastic:changeme --insecure "https://localhost:9200/_security/user/test-uptime" -H 'kbn-xsrf: true' -H 'Content-Type: application/json' -d'
{
  "password" : "testuptime",
  "roles" : [ "uptime_role" ],
  "full_name" : "Test User",
  "email" : "test@test.test"
}
```
4. Log in as a superuser using the `elastic` user.
5. Navigate to monitor management. Confirm the `Enable` toggle is shown. If you're on a fresh es instance, the enable toggle should be off. If you are on an instance that already has monitors, the enable toggle with be on.
6. Click the enable toggle to enable or disable (depending on the state of your cluster). Confirm that the success toast appears. When enabling, the Add Monitor button should be enabled. When disabling, the Add Monitor button should be disabled.
7. Ensure that monitor management is disabled. Log out and log in as the `test-uptime` user.
8. Confirm that the Add Monitor button is enabled for `test-uptime`. Hover over the button to view a tooltip.
9. Log back in as `elastic`. Enable monitor management. Log out and log in as the `test-uptime` user.
10. Confirm that the Add Monitor button is enabled.

### Testing sync
1. Connect to the dev instance of the synthetics service, either by running the synthetics service locally or connecting via `xpack.uptime.service.manifestUrl`. Ensure your es instance can be accessed, either by using `oblt-cli` or exposing your local es instance via ngrok, and ensure `xpack.uptime.service.hosts` is defined.
2. Log in as the `elastic` user. Enable monitor management and create a monitor. Ensure that the monitor is indexed appropriately on the overview page.
3. Navigate back to monitor management. Disable monitor management. Ensure that monitors stop running immediately by navigating to Uptime Overview and ensuring that documents have stopped indexing. 

### Checklist

Delete any items that are not applicable to this PR.
